src/nepiada/env/nepiada.py

```diff
@@ -111,6 +115,61 @@ def close(self):
         """
         pass
 
+    def get_observations(self):
+        """
+        The n^2 observation structure returned below are the coordinates of each agents that each agent can directly observe
+        observations[i][j] is the location that drone i sees drone j at
+        """
+        observations = {agent: None for agent in self.agents}
+        for agent_name in self.agents:
+            observation = {}
+            for observed_agent_name in self.agents:
+                observed_agent = self.world.get_agent(observed_agent_name)
+                if observed_agent_name in self.world.graph.obs[agent_name]:
+                    observation[observed_agent_name] = (observed_agent.p_pos[0], observed_agent.p_pos[1])
+                else:
+                    observation[observed_agent_name] = None # Cannot be observed
+            observations[agent_name] = observation
+        return observations
+    
+    def get_all_messages(self):
+        """
+        The n^3 message structure returned below are the coordinates that each drone receives from a drone about another drone
+        observations[i][j][k] is the location that drone i is told by drone k where drone k is
+        """
+        incoming_all_messages = {}
+        for agent_name in self.agents:
+            beliefs = self.world.get_agent(agent_name).beliefs
+            observation = self.observations[agent_name]
+
+            incoming_agent_messages = {}
+
+            for target_agent_name in self.agents:
+
+                incoming_communcation_messages = {}
+
+                if not observation[target_agent_name]: # Must estimate where the agent is via communication
+                    for helpful_agent in self.world.graph.comm[agent_name]:  
+                        helpful_beliefs = self.noise.add_noise(self.world.get_agent(helpful_agent).beliefs)
+                        if helpful_beliefs[target_agent_name]:
+                            incoming_communcation_messages[helpful_agent] = (helpful_beliefs[target_agent_name][0], helpful_beliefs[target_agent_name][1])
+                
+                incoming_agent_messages[target_agent_name] = incoming_communcation_messages
+
+            incoming_all_messages[agent_name] = incoming_agent_messages
+        return incoming_agent_messages
+    
+    def update_beliefs(self):
```

Comment:
Do you think it would be better to let the algorithm complete all of the beliefs updates? Including observation? It may be a bit confusing having the beliefs that can be observed be updated by the environment and beliefs that are not being updated by the algorithm

Reply:
@pandyah5

Reply:
Making beliefs update external to the env is a good idea, but not the observation. The observation graph should follow a single criterion: observation radius - this should not be modifiable externally.

Reply:
Right now, the communication messages part depends on the existing beliefs of the other agents, so what can be observed should be directly given, but we can think of a workaround.

Reply:
I don't mean modifying the observation graph. We would still return the observation graph. I mean letting agents update their beliefs based on the observations + communications themselves, rather than the environment updating the beliefs, since we return the observations + communications to the algorithm.

Reply:
Oh, I get it now. Yes I remember we discussed this, it should be internal to the agent class like:

class Agent:
    def update(self):
          # here

# In env step function
agent.update()

Reply:
@Thanoshan  @pandyah5  I created another function to just initialize all the beliefs to an empty value, it should work for our use cases.

Reply:
Can you explain how it would address the point of agent updating their beliefs internally?

Reply:
Yeah this doesn't address the issue

Reply:
> Oh, I get it now. Yes I remember we discussed this, it should be internal to the agent class like:
> 
> class Agent: def update(self): # here
> 
> #In env step function agent.update()

It doesn't really have to be part of the Agent class, the algorithm would just be responsible for taking the observations and communications, and updating the agent's beliefs accordingly. The agent class could have an update_beliefs method that just takes a dictionary and replaces the beliefs dictionary if anything (like a setter)

Reply:
That works too

Reply:
update_beliefs is unused, i can delete it now if you want. The environment no longer auto sets any beliefs

Reply:
Maybe move this method to launch.py for now. Though it would be changed so that the algorithm retrieves the current beliefs for the agent, then updates the entries in the dictionary corresponding to the agents that it observes.

Reply:
That is part of the algorithm so we'll decide how to do that next?

Reply:
If you'd rather separate the PR for it, add a comment to the update_beliefs method saying that it's deprecated

Reply:
wait the code already uses the observations